lib/Test/Mockify/Base.pm

```diff
@@ -0,0 +1,249 @@
+package Test::Mockify::Base;
+use strict;
+use warnings;
+
+use Test::Mockify::TypeTests qw ( IsString IsArrayReference );
+use Test::Mockify::Tools qw ( LoadPackage ExistsMethod Error );
+use Test::Mockify::MethodCallCounter;
+use Test::Mockify::Method;
+use Test::Mockify::MethodSpy;
+use Test::Mockify::CompatibilityTools qw ( MigrateOldMatchers );
+
+use Scalar::Util qw( blessed );
+use Test::MockObject::Extends;
+
+sub new {
+    my ($class, $FakeModulePath, $aFakeParams) = @_;
+
+    LoadPackage($FakeModulePath);
+
+    my $self = bless {}, $class;
+    my $FakeClass = $self->_getFakeClass($FakeModulePath, $aFakeParams);
+    $self->_mockedModulePath($FakeModulePath);
+    $self->_mockedSelf(Test::MockObject::Extends->new($FakeClass));
+    $self->_initMockedModule();
+
+    return $self;
+}
+
+sub mock {
+    my $self = shift;
+    my @Parameters = @_;
+
+    my $ParameterAmount = scalar @Parameters;
+    if($ParameterAmount == 1 && IsString($Parameters[0]) ){
+        return $self->_addMockWithMethod($Parameters[0]);
+    }
+    if($ParameterAmount == 2){
+        my ( $MethodName, $ReturnValueOrFunctionPointer ) = @Parameters;
+        if( ref($ReturnValueOrFunctionPointer) eq 'CODE' ){
+            $self->addMock($MethodName, $ReturnValueOrFunctionPointer);
+        }else{
+            $self->addMockWithReturnValue($MethodName, $ReturnValueOrFunctionPointer);
+        }
+    }
+    if($ParameterAmount == 3){
+        my ( $MethodName, $ReturnValue, $aParameterTypes ) = @_;
+        $self->addMockWithReturnValueAndParameterCheck($MethodName, $ReturnValue, $aParameterTypes);
+    }
+    return;
+}
+
+sub spy {
+    my $self = shift;
+    my ($MethodName) = @_;
+    my $PointerOriginalMethod = \&{$self->_mockedModulePath().'::'.$MethodName};
+    my $mockedSelf = $self->_mockedSelf();
+    #In order to have the current object available in the parameter list, it has to be injected here.
+    return $self->_addMockWithMethodSpy($MethodName, sub {
+            return $PointerOriginalMethod->($mockedSelf, @_);
+        });
+}
+
+sub getVerifier {
```

Comment:
As you probably read above, that function is mainly for the Injector. I'm not altogether happy with the way the getMockObject/getVerifier part turned out. You may have noticed that getMockObject simply calls getVerifier. At one point I had getMockObject and getVerifier on Mockify and Injector, respectively, and they each had the same implementation:

```perl
my $self = shift;
return $self->getMockObject();
```

That felt slightly wrong too. I think if we work toward the solution I proposed to your comment about why did I create Base, this will be resolved.

Reply:
The whole not going out of scope thing was because a direct reference to Mockify's $self was captured inside of several anonymous subs. This prevented the Mockify instance from ever being released. You can review this commit for how I solved it: https://github.com/dbucky/Mockify/commit/2924c49178c9d2594584ddb2ad25dea0368a27c5.

Reply:
wow! this fixes my scope problems. THX I didn't found it.